AdaptiveTimeStepping: allow for more flexible error computation in the future.

dr-robertk opened the pull request:
This PR revises the TimeStepControl interface such that the computation of the 
error || u^n+ - u^n || / || u^n+1 || is forwarded to the Model class. This allows for more flexible and model dependent implementation. A corresponding PR in opm-autodiff is coming.

atgeirr commented on opm/core/simulator/AdaptiveTimeStepping_impl.hpp:
```diff
@@ -31,11 +31,35 @@
+            {}
+
+            /// return || u^n+1 - u^n || / || u^n+1 ||
+            double timeError() const
```
I think timeError() is a strange name for this, would not relativeChange() or something like that be better? I interpret timeError as being an error in a time variable or something like that.

atgeirr commented on opm/core/simulator/TimeStepControl.hpp:
```diff
@@ -82,60 +82,18 @@ namespace Opm
         const bool verbose_;
     private:
-        const boost::any parallel_information_;
+    //    const boost::any parallel_information_;
```
Should be removed entirely.

atgeirr commented on opm/core/simulator/TimeStepControlInterface.hpp:
```diff
@@ -19,31 +19,45 @@
     ///
-    ///  TimeStepControlInterface 
-    /// 
+    ///  TimeStepControlInterface
```
Wrong name?

atgeirr commented on opm/core/simulator/AdaptiveTimeStepping_impl.hpp:
```diff
@@ -172,9 +193,13 @@ namespace Opm {
                 // compute new time step estimate
                 double dtEstimate =
-                    timeStepControl_->computeTimeStepSize( dt, linearIterations, state );
+                    timeStepControl_->computeTimeStepSize( dt, linearIterations, timeError );
```
Since the "time error" is always a double, would it not be simpler if there was no SolutionTimeErrorSolver* classes, that the SolutionTimeErrorSolverWrapper timeError() method content was a small method in this class (AdaptiveTimeStepping) and the computeTimeStepSize() methods took a double instead of the STESInterface?

Then this would look like:
```
const double timeError = computeTimeError(solver, last_state, state); // method call
timeStepControl_->computeTimeStepSize(dt, linearIterations, timeError); // passing a double
```
I am not against using classes and inheritance, but in this case I do not see the benefit.

dr-robertk commented:
I have addressed the concerns. timeError has been renamed to relativeChange. 
It is not a good idea to pass a simple double, because then the relativeChange has to be computed in any case, whereas with my solution it's only computed when needed.

bska commented on opm/core/simulator/AdaptiveTimeStepping_impl.hpp:
```diff
@@ -31,11 +31,35 @@
+    namespace detail
+    {
+        template <class Solver, class State>
+        class SolutionTimeErrorSolverWrapper : public SolutionTimeErrorInterface
```
What is the `SolutionTimeErrorInterface`?  It does not appear to exist in any OPM source file.

atgeirr replied:
I do not know why I did not encounter a build error when I tested this, I do get one now in any case! Will fix.

bska replied:
> I do not know why I did not encounter a build error when I tested this,

Did you build-test opm-core in isolation?  I only get a build-failure in downstream module opm-autodiff.

atgeirr replied:
Perhaps, but I also built in opm-autodiff (I think). @dr-robertk has suggested that these parts should be moved to opm-autodiff in any case, and I support that. Fix is easy, coming soon.
